File: apps/web/lib/ielts.ts
import { apiFetch } from "@/lib/api";

export type IeltsSkill = "reading" | "writing" | "listening" | "speaking";
export type ComprehensionKind = "reading" | "listening";

export interface IeltsReward {
  xp_gained: number;
  total_xp: number;
  level: number;
  leveled_up: boolean;
}

export interface IeltsHistoryItem {
  skill: string;
  band: number;
  correct: number | null; // Reading/Listening only
  total: number | null;
  created_at: string;
}

export interface IeltsOverview {
  best_bands: Record<string, number>;
  recent: IeltsHistoryItem[]; // newest first
  enabled: boolean;
}

export interface WritingChartSeries {
  name: string;
  values: number[];
}

export interface WritingPieSlice {
  name: string;
  value: number;
}

export interface WritingPieChart {
  title: string;
  slices: WritingPieSlice[];
}

export interface WritingMapFeature {
  name: string;
  kind: "water" | "road" | "building" | "green";
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface WritingMap {
  title: string;
  features: WritingMapFeature[];
}

export interface WritingSecondaryChart {
  title: string;
  y_label?: string;
  categories: string[];
  series: WritingChartSeries[];
}

export interface WritingTaskVisual {
  kind: "line" | "bar" | "table" | "bar-pie" | "pie-pair" | "process" | "image" | "map-pair" | "bar-line";
  title: string;
  y_label?: string;
  categories: string[];
  series: WritingChartSeries[];
  pie?: WritingPieSlice[];
  pies?: WritingPieChart[];
  maps?: WritingMap[];
  secondary?: WritingSecondaryChart;
  // "image" only: a real diagram or chart under /public, shown as-is instead of
  // being redrawn from categories/series the way the data-driven kinds are.
  image?: string;
  image_alt?: string;
}

export interface WritingTask {
  title: string;
  prompt: string;
  visual?: WritingTaskVisual | null;
}

export interface IeltsQuestion {
  prompt: string;
  options: string[];
}

export interface GeneratedTest {
  test_id: string;
  title: string;
  body: string; // reading passage or listening script
  questions: IeltsQuestion[];
}

export interface GradeResult {
  correct: number;
  total: number;
  band: number;
  /** Short practice sets cannot resolve a single band — show a range instead. */
  approximate: boolean;
  answers: number[];
  /** Why each answer is right, in question order. Empty for bank items, which
   *  have hand-written passages but no explanations yet. */
  explanations: string[];
  reward: IeltsReward;
}

export interface WritingCriterion {
  band: number;
  comment: string;
}

export interface WritingError {
  quote: string; // exact fragment from the essay
  fix: string; // corrected fragment
  note: string; // one-sentence explanation
  type: "grammar" | "vocabulary" | "spelling" | "punctuation" | "style";
}

export type WritingFeedbackStatus = "good" | "improve" | "error";

export type WritingFeedbackCategory =
  | "grammar"
  | "vocabulary"
  | "collocation"
  | "articles"
  | "prepositions"
  | "word_form"
  | "tense"
  | "subject_verb_agreement"
  | "sentence_structure"
  | "punctuation"
  | "cohesion"
  | "logic"
  | "style"
  | "spelling";

export interface WritingSentenceFeedback {
  sentence_number: number;
  sentence: string;
  highlight: string;
  status: WritingFeedbackStatus;
  category: WritingFeedbackCategory;
  explanation: string;
  use_instead: string;
  why: string;
}

export interface WritingGoodPoint {
  title: string;
  evidence: string;
  explanation: string;
}

export interface WritingImprovementArea {
  title: string;
  evidence: string;
  action: string;
}

export interface WritingLanguageUpgrade {
  used: string;
  use_instead: string;
  why: string;
}

export interface WritingRepetition {
  word: string;
  frequency: number;
  problem: string;
  alternatives: string[];
}

export interface WritingObservation {
  quote: string;
  explanation: string;
}

export interface WritingCohesionAnalysis {
  strengths: WritingObservation[];
  issues: WritingObservation[];
  opportunities: string[];
}

export interface WritingGrammarProfile {
  strengths: WritingObservation[];
  weaknesses: WritingObservation[];
}

export interface WritingBandPlan {
  current_band: number;
  target_band: number;
  actions: string[];
}

export interface WritingAnalysis {
  sentence_feedback: WritingSentenceFeedback[];
  good_points: WritingGoodPoint[];
  areas_to_improve: WritingImprovementArea[];
  language_upgrades: WritingLanguageUpgrade[];
  repetitions: WritingRepetition[];
  cohesion: WritingCohesionAnalysis;
  grammar_profile: WritingGrammarProfile;
  band_plan: WritingBandPlan;
  next_steps: string[];
}

export interface WritingScore {
  band_overall: number;
  task: WritingCriterion;
  coherence: WritingCriterion;
  lexical: WritingCriterion;
  grammar: WritingCriterion;
  errors: WritingError[];
  strengths: string[];
  feedback: string;
  improved: string; // full band-8 model rewrite
  analysis: WritingAnalysis;
  reward: IeltsReward;
}

export interface BankItem {
  id: string;
  title: string;
  band: number;
  question_count: number;
  word_count: number;
  done: boolean;
}

export const ieltsApi = {
  overview: () => apiFetch<IeltsOverview>("/ielts/overview", { auth: true }),

  bank: (kind: ComprehensionKind) =>
    apiFetch<BankItem[]>(`/ielts/${kind}/bank`, { auth: true }),

  bankStart: (kind: ComprehensionKind, itemId: string) =>
    apiFetch<GeneratedTest>(`/ielts/${kind}/bank/${itemId}/start`, {
      method: "POST",
      auth: true,
    }),

  writingTasks: () =>
    apiFetch<Record<string, WritingTask[]>>("/ielts/writing/tasks", { auth: true }),

  scoreWriting: (taskType: string, prompt: string, essay: string, lang = "en", mockSessionId?: string) =>
    apiFetch<WritingScore>("/ielts/writing/score", {
      method: "POST",
      body: { task_type: taskType, prompt, essay, lang, mock_session_id: mockSessionId },
      auth: true,
      // A grounded, sentence-level report is intentionally a longer model
      // call than ordinary API actions. Match the server-side AI allowance so
      // the browser does not abandon a healthy report after the global 15s.
      timeoutMs: 75_000,
    }),

  generate: (kind: ComprehensionKind, band = 6) =>
    apiFetch<GeneratedTest>(`/ielts/${kind}/generate`, {
      method: "POST",
      body: { band },
      auth: true,
    }),

  submit: (kind: ComprehensionKind, testId: string, answers: number[], mockSessionId?: string) =>
    apiFetch<GradeResult>(`/ielts/${kind}/submit`, {
      method: "POST",
      body: { test_id: testId, answers, mock_session_id: mockSessionId },
      auth: true,
    }),
};

export const BAND_COLOR = (band: number): string => {
  if (band >= 7.5) return "text-accent-text";
  if (band >= 6.5) return "text-brand-500";
  if (band >= 5.5) return "text-brand-400";
  return "text-brand-600";
};

File: apps/api/app/schemas/ielts.py
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class RewardOut(BaseModel):
    xp_gained: int
    total_xp: int
    level: int
    leveled_up: bool


class WritingTask(BaseModel):
    title: str
    prompt: str
    # Task 1 prompts may include a compact chart description for the web client.
    # Keeping the data alongside the prompt lets learners see the visual before
    # they start writing, without fetching a separate image asset.
    visual: Optional[Dict[str, Any]] = None


class HistoryItemOut(BaseModel):
    skill: str
    band: float
    correct: Optional[int] = None  # Reading/Listening only
    total: Optional[int] = None
    created_at: datetime


class OverviewOut(BaseModel):
    best_bands: Dict[str, float]  # skill -> best band
    recent: List[HistoryItemOut]  # newest first
    enabled: bool


class QuestionOut(BaseModel):
    prompt: str
    options: List[str]


class BankItemOut(BaseModel):
    id: str
    title: str
    band: float  # approximate difficulty, for sorting/labelling
    question_count: int
    word_count: int
    done: bool


class GenerateRequest(BaseModel):
    band: float = Field(default=6.0, ge=4.0, le=9.0)


class GeneratedTestOut(BaseModel):
    test_id: UUID
    title: str
    body: str  # reading passage OR listening script (spoken by the browser)
    questions: List[QuestionOut]


class SubmitRequest(BaseModel):
    test_id: UUID
    answers: List[int]
    # Set when this submission is one leg of an IELTS Full Mock exam, so the
    # resulting IeltsResult row can be tied back to that session.
    mock_session_id: Optional[UUID] = None


class GradeOut(BaseModel):
    correct: int
    total: int
    band: float
    # Short practice sets cannot resolve a single band; the client shows a range.
    approximate: bool = False
    answers: List[int]  # correct indices, revealed after grading
    explanations: List[str] = []  # why each answer is right, same order
    reward: RewardOut


class WritingScoreRequest(BaseModel):
    task_type: str = Field(pattern="^(task1|task2)$")
    prompt: str = Field(min_length=10, max_length=1200)
    essay: str = Field(min_length=20, max_length=6000)
    lang: str = Field(default="en", pattern="^(uz|ru|en)$")  # feedback language
    mock_session_id: Optional[UUID] = None


class CriterionOut(BaseModel):
    band: float
    comment: str


class WritingErrorOut(BaseModel):
    quote: str  # exact fragment from the essay
    fix: str  # corrected fragment
    note: str  # one-sentence explanation in the requested language
    type: str  # grammar|vocabulary|spelling|punctuation|style


WritingFeedbackStatus = Literal["good", "improve", "error"]
WritingFeedbackCategory = Literal[
    "grammar",
    "vocabulary",
    "collocation",
    "articles",
    "prepositions",
    "word_form",
    "tense",
    "subject_verb_agreement",
    "sentence_structure",
    "punctuation",
    "cohesion",
    "logic",
    "style",
    "spelling",
]


class WritingSentenceFeedbackOut(BaseModel):
    sentence_number: int
    sentence: str
    highlight: str
    status: WritingFeedbackStatus
    category: WritingFeedbackCategory
    explanation: str
    use_instead: str
    why: str


class WritingGoodPointOut(BaseModel):
    title: str
    evidence: str
    explanation: str


class WritingAreaToImproveOut(BaseModel):
    title: str
    evidence: str
    action: str


class WritingLanguageUpgradeOut(BaseModel):
    used: str
    use_instead: str
    why: str


class WritingRepetitionOut(BaseModel):
    word: str
    frequency: int
    problem: str
    alternatives: List[str]


class WritingQuotedAnalysisPointOut(BaseModel):
    quote: str
    explanation: str


class WritingCohesionOut(BaseModel):
    strengths: List[WritingQuotedAnalysisPointOut]
    issues: List[WritingQuotedAnalysisPointOut]
    opportunities: List[str]


class WritingGrammarProfileOut(BaseModel):
    strengths: List[WritingQuotedAnalysisPointOut]
    weaknesses: List[WritingQuotedAnalysisPointOut]


class WritingBandPlanOut(BaseModel):
    current_band: float
    target_band: float
    actions: List[str]


class WritingAnalysisOut(BaseModel):
    sentence_feedback: List[WritingSentenceFeedbackOut]
    good_points: List[WritingGoodPointOut]
    areas_to_improve: List[WritingAreaToImproveOut]
    language_upgrades: List[WritingLanguageUpgradeOut]
    repetitions: List[WritingRepetitionOut]
    cohesion: WritingCohesionOut
    grammar_profile: WritingGrammarProfileOut
    band_plan: WritingBandPlanOut
    next_steps: List[str]


class WritingQuotaOut(BaseModel):
    """What the essay composer shows before the learner starts writing, so a
    limit is never discovered by hitting it."""

    # "week" on the free plan, "day" on a paid one — the two tiers are capped
    # over different windows.
    period: Literal["week", "day"]
    limit: int
    used: int
    remaining: int
    premium: bool


class MockQuotaOut(BaseModel):
    free_attempt_available: bool
    premium: bool
    coin_cost: int
    coin_balance: int


class QueuedJobOut(BaseModel):
    """202 response for work handed to the background worker; poll
    GET /jobs/{job_id} for the result."""

    job_id: UUID


def writing_score_out(score) -> "WritingScoreOut":
    """Builds the response from a services.ielts WritingScore.

    Lives here rather than in the route because the queue worker produces the
    very same payload out of band — see services/job_handlers.py."""
    return WritingScoreOut(
        band_overall=score.band_overall,
        task=CriterionOut(band=score.task.band, comment=score.task.comment),
        coherence=CriterionOut(band=score.coherence.band, comment=score.coherence.comment),
        lexical=CriterionOut(band=score.lexical.band, comment=score.lexical.comment),
        grammar=CriterionOut(band=score.grammar.band, comment=score.grammar.comment),
        errors=[
            WritingErrorOut(quote=e.quote, fix=e.fix, note=e.note, type=e.type)
            for e in score.errors
        ],
        strengths=score.strengths,
        feedback=score.feedback,
        improved=score.improved,
        analysis=WritingAnalysisOut(**score.analysis),
        reward=RewardOut(
            xp_gained=score.reward.xp_gained,
            total_xp=score.reward.total_xp,
            level=score.reward.level,
            leveled_up=score.reward.leveled_up,
        ),
    )


class WritingScoreOut(BaseModel):
    band_overall: float
    task: CriterionOut
    coherence: CriterionOut
    lexical: CriterionOut
    grammar: CriterionOut
    errors: List[WritingErrorOut]
    strengths: List[str]
    feedback: str
    improved: str  # full band-8 model rewrite
    analysis: WritingAnalysisOut
    reward: RewardOut
